File: apps/fsm-core-db-drizzle/src/queue.ts
import { sql } from "drizzle-orm";

import type { Database } from "./database.types.ts";
import type { DBDeps } from "./custom-type.ts";

import { QUEUE_SCHEMA } from "./const.ts";


const DELETE_QUEUE_FN = `${QUEUE_SCHEMA}.delete`;
const ARCHIVE_QUEUE_FN = `${QUEUE_SCHEMA}.archive`;
const LIST_QUEUES_FN = `${QUEUE_SCHEMA}.list_queues`;

export async function readMessage(
  deps: DBDeps,
  queueName: string,
  vt: number,
): Promise<Database["pgmq"]["CompositeTypes"]["message_record"][]> {
  try {
    const READ_QUEUE_FN = `${QUEUE_SCHEMA}.read`;
    const qty = 1; // Read one message at a time for processing
    const query = sql`
      SELECT * FROM ${sql.raw(READ_QUEUE_FN)}(
        ${queueName}::text,
        ${vt}::integer,
        ${qty}::integer
      );
    `;
    const res = await deps.db.execute(query);
    return res.rows;
  } catch (err) {
    console.error("Error in readMessage:", err);
    return [];
  }
}

export async function deleteMessage(
  deps: DBDeps,
  queueName: string,
  msgId: number,
): Promise<void> {
  try {
    const query = sql`
      SELECT * FROM ${sql.raw(DELETE_QUEUE_FN)}(
        ${queueName}::text,
        ${msgId}::bigint
      );
    `;
    await deps.db.execute(query);
  } catch (err) {
    console.error("Error in deleteMessage:", err);
  }
}

export async function archiveMessage(
  deps: DBDeps,
  queueName: string,
  msgId: number,
): Promise<void> {
  try {
    const query = sql`
      SELECT * FROM ${sql.raw(ARCHIVE_QUEUE_FN)}(
        ${queueName}::text,
        ${msgId}::bigint
      );
    `;
    await deps.db.execute(query);
  } catch (err) {
    console.error("Error in archiveMessage:", err);
  }
}


/**
 * Checks whether a PGMQ queue with the given name exists in the database.
 * Uses the `${QUEUE_SCHEMA}.list_queues()` wrapper when available.
 * @param deps - DBDeps containing either supabase or drizzle client
 * @param queueName - The PGMQ queue name to check
 * @returns Promise<boolean>
 */
export async function pgmqQueueExists(
  deps: DBDeps,
  queueName: string,
): Promise<boolean> {
  if (!queueName) return false;
  try {
    const query = sql`
      SELECT * FROM ${sql.raw(LIST_QUEUES_FN)}();
    `;
    const res = await deps.db.execute(query);
    const rows = res.rows ?? [];
    return rows.some((r: any) =>
      r?.name === queueName || r?.queue_name === queueName
    );
  } catch (err) {
    console.error("Error in pgmqQueueExists:", err);
    return false;
  }
}

File: apps/fsm-core-py-db/src/fsm_core_db/queue.py
import asyncpg

from .constants import QUEUE_SCHEMA

READ_FN = f"{QUEUE_SCHEMA}.read"
DELETE_FN = f"{QUEUE_SCHEMA}.delete"
ARCHIVE_FN = f"{QUEUE_SCHEMA}.archive"
LIST_QUEUES_FN = f"{QUEUE_SCHEMA}.list_queues"


async def read_message(
    pool: asyncpg.Pool, queue_name: str, vt: int, qty: int = 1
) -> list[dict]:
    """
    Read up to `qty` messages from a pgmq queue with a visibility timeout of `vt` seconds.
    Mirrors readMessage() in queue.ts.
    """
    sql = f"SELECT * FROM {READ_FN}($1::text, $2::integer, $3::integer);"
    async with pool.acquire() as conn:
        rows = await conn.fetch(sql, queue_name, vt, qty)
    return [dict(r) for r in rows]


async def delete_message(pool: asyncpg.Pool, queue_name: str, msg_id: int) -> None:
    """
    Permanently delete a message from a pgmq queue.
    Mirrors deleteMessage() in queue.ts.
    """
    sql = f"SELECT * FROM {DELETE_FN}($1::text, $2::bigint);"
    async with pool.acquire() as conn:
        await conn.execute(sql, queue_name, msg_id)


async def archive_message(pool: asyncpg.Pool, queue_name: str, msg_id: int) -> None:
    """
    Move a message to the pgmq archive table.
    Mirrors archiveMessage() in queue.ts.
    """
    sql = f"SELECT * FROM {ARCHIVE_FN}($1::text, $2::bigint);"
    async with pool.acquire() as conn:
        await conn.execute(sql, queue_name, msg_id)


async def pgmq_queue_exists(pool: asyncpg.Pool, queue_name: str) -> bool:
    """
    Return True if a pgmq queue with the given name exists.
    Mirrors pgmqQueueExists() in queue.ts.
    """
    if not queue_name:
        return False
    sql = f"SELECT * FROM {LIST_QUEUES_FN}();"
    async with pool.acquire() as conn:
        rows = await conn.fetch(sql)
    return any(
        r.get("name") == queue_name or r.get("queue_name") == queue_name
        for r in rows
    )
